### PR TITLE
Fix recognition of "between now and date" as daterange

### PR DESCRIPTION
**Pull request by songwenhao1**
Can't recognize "between now and date" daterange:
Query: between now and November 15th
wrong answer: two text: "now" "November 15th"
right answer: a daterange text: "between now and November 15th"
Supported in .NET,  java, python

 Javascript implementation will come in a future PR

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 1]
```diff
@@ -1257,67 +1262,131 @@ private DateTimeResolutionResult ParseYear(string text, DateObject referenceDate
...
+            ret.Timex = $"({pr1.TimexStr},{pr2.TimexStr},P{(futureEnd - futureBegin).TotalDays}D)";
+            ret.FutureValue = new Tuple<DateObject, DateObject>(futureBegin, futureEnd);
+            ret.PastValue = new Tuple<DateObject, DateObject>(pastBegin, pastEnd);
+            ret.Success = true;
+
+            return ret;
+        }
+
+        // Parse entities that are made up by two time points with now
+        private DateTimeResolutionResult MergeTwoTimePointsWithNow(string text, DateObject referenceDate)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Why do we need two methods? Also, are the other changes just line ending changes? If so, please revert them to make review easier.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 2]
```diff
@@ -1257,67 +1262,131 @@ private DateTimeResolutionResult ParseYear(string text, DateObject referenceDate
...
         {
             var ret = new DateTimeResolutionResult();
-
+            var nowMatches = this.config.NowRegex.Matches(text);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Didn't extract already tag "now"? Why do we need to re-extract in Parser?

File: .NET/Microsoft.Recognizers.Text.sln [hunk 3]
```diff
@@ -4,6 +4,9 @@ Microsoft Visual Studio Solution File, Format Version 12.00
 VisualStudioVersion = 15.0.27130.2026
 MinimumVisualStudioVersion = 10.0.40219.1
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Tests", "Tests", "{ED7B6456-AB0A-48CE-8F85-711FE87F09C2}"
+	ProjectSection(SolutionItems) = preProject
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.sln]: Why is this change necessary?

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/extractors/BaseDatePeriodExtractor.java [hunk 4]
```diff
@@ -21,6 +21,8 @@
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
+
```
Comment (tellarin) [inline Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/extractors/BaseDatePeriodExtractor.java]: Nit: Please, be careful to not add unnecessary blank lines like these.

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/extractors/BaseDatePeriodExtractor.java [hunk 5]
```diff
@@ -95,6 +97,15 @@ public BaseDatePeriodExtractor(IDatePeriodExtractorConfiguration config) {
...
 
+        // Handle "now"
+        Match[] matches = RegExpUtility.getMatches(this.config.getNowRegex(), input);
+        if (matches.length != 0) {
+            for (Match match : matches) {
+                ers.add(new ExtractResult(match.index, match.length, match.value, Constants.SYS_DATETIME_DATE));
+            }
+
+            ers.sort(Comparator.comparingInt(arg -> arg.getStart()));
+        }
```
Comment (tellarin) [inline Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/extractors/BaseDatePeriodExtractor.java]: Nit: Add a blank line between closing brace and the return.

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDateParser.java [hunk 6]
```diff
@@ -177,8 +177,7 @@ private DateTimeResolutionResult parseImplicitDate(String text, LocalDateTime re
 
         if (exactMatch.getSuccess()) {
             int swift = getSwiftDay(exactMatch.getMatch().get().value);
-
-            LocalDateTime value = referenceDate.plusDays(swift);
+            LocalDateTime value = referenceDate.toLocalDate().atStartOfDay().plusDays(swift);
```
Comment (tellarin) [inline Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDateParser.java]: Is this really the same behaviour as in .NET? Any usage of toLocal() always make me nervous of causing timezone issues.

File: Specs/DateTime/English/DatePeriodExtractor.json [hunk 7]
```diff
@@ -3717,5 +3717,16 @@
         "Length": 17
       }
     ]
+  },
+  {
+    "Input": "I will do my work between now and November 15th",
```
Comment (tellarin) [inline Specs/DateTime/English/DatePeriodExtractor.json]: Can you also add a case for "between Jan 22 and now"?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 1 above)
Comment (songwenhao1) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: In the first method, we will re_extract "date" from  "text", "now" is a Date time, so it would not be extracted, could not solve the issue

File: Specs/DateTime/English/DatePeriodParser.json [hunk 8]
```diff
@@ -5595,5 +5595,31 @@
...
+          "Timex": "(2019-04-23,2019-11-15,P206D)",
+          "FutureResolution": {
+            "startDate": "2019-04-23",
+            "endDate": "2019-11-15"
+          },
+          "PastResolution": {
+            "startDate": "2019-04-23",
+            "endDate": "2019-11-15"
+          }
+
```
Comment (STayinloves) [inline Specs/DateTime/English/DatePeriodParser.json]: Remove this blank line

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDateParser.java (hunk 6 above)
Comment (songwenhao1) [reply Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDateParser.java]: toLocalDate() just return date information.

public LocalDate toLocalDate() {
    return date;
} 
i find this also used in resolve End Of Day

    private DateTimeResolutionResult resolveEndOfDay(String timexPrefix, LocalDateTime futureDate, LocalDateTime pastDate) {
        String timex = String.format("%sT23:59:59", timexPrefix);
        LocalDateTime futureValue = LocalDateTime.of(futureDate.toLocalDate(), LocalTime.MIDNIGHT).plusDays(1).minusSeconds(1);
        LocalDateTime pastValue = LocalDateTime.of(pastDate.toLocalDate(), LocalTime.MIDNIGHT).plusDays(1).minusSeconds(1);

        DateTimeResolutionResult result = new DateTimeResolutionResult();
        result.setTimex(timex);
        result.setFutureValue(futureValue);
        result.setPastValue(pastValue);
        result.setSuccess(true);

        return result;
    }

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 1 above)
Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: There is also a change in the extractor extracting "now" in this PR, no? Why do we need it again parsing?

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: And it could be part of the same parsing method as a special case, couldn't it?

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDateParser.java (hunk 6 above)
Comment (tellarin) [reply Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDateParser.java]: OK.

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDatePeriodParser.java [hunk 9]
```diff
@@ -939,42 +940,57 @@ private DateTimeResolutionResult mergeTwoTimePoints(String text, LocalDateTime r
...
+            }
 
-        DateTimeParseResult pr1 = this.config.getDateParser().parse(er.get(0), referenceDate);
-        DateTimeParseResult pr2 = this.config.getDateParser().parse(er.get(1), referenceDate);
+            DateContext dateContext = getYearContext(er.get(0).getText(), er.get(1).getText(), text);
 
-        if (pr1.getValue() == null || pr2.getValue() == null) {
-            return ret;
-        }
+            pr1 = this.config.getDateParser().parse(er.get(0), referenceDate);
```
Comment (STayinloves) [inline Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDatePeriodParser.java]: We should only `getDateParser` once to minimize time&memory cost.

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDateParser.java [hunk 10]
```diff
@@ -177,8 +177,8 @@ private DateTimeResolutionResult parseImplicitDate(String text, LocalDateTime re
 
         if (exactMatch.getSuccess()) {
             int swift = getSwiftDay(exactMatch.getMatch().get().value);
-
-            LocalDateTime value = referenceDate.plusDays(swift);
+
```
Comment (STayinloves) [inline Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDateParser.java]: Remove trailing spaces.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 11]
```diff
@@ -1325,6 +1345,39 @@ private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject refe
             return ret;
         }
 
+        // Parse entities that are made up by two time points with now
+        private List<DateTimeParseResult> GetPrWithNow(MatchCollection matches, ExtractResult er, DateObject referenceDate)
+        {
+            var pr = new List<DateTimeParseResult>();
+            pr.Add(this.config.DateParser.Parse(er, referenceDate));
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: This method should be only for the "now" code. The other entity (the date result) should not be extracted here.
Also, why call Parse() and not Extract like in the original code?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 12]
```diff
@@ -1262,43 +1262,63 @@ private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject refe
             var ret = new DateTimeResolutionResult();
 
             var er = this.config.DateExtractor.Extract(text, referenceDate);
-            if (er.Count < 2)
-            {
-                er = this.config.DateExtractor.Extract(this.config.TokenBeforeDate + text, referenceDate);
-                if (er.Count < 2)
-                {
-                    return ret;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Shouldn't the attempt to extract "now" be here, and if nothing is extracted, then return ret?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 13]
```diff
@@ -1262,43 +1262,63 @@ private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject refe
...
-            var match = this.config.WeekWithWeekDayRangeRegex.Match(text);
-            string weekPrefix = null;
-            if (match.Success)
-            {
-                weekPrefix = match.Groups["week"].ToString();
+            var pr1 = new DateTimeParseResult();
+            var pr2 = new DateTimeParseResult();
+
+            // Handle with "now"
+            var nowMatches = this.config.NowRegex.Matches(text);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: We shouldn't have to even do this, if the previous Extract already returned two entities.

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: And everything related to "now", including this regex match should go into the method.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 14]
```diff
@@ -1325,6 +1345,39 @@ private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject refe
             return ret;
         }
 
+        // Parse entities that are made up by two time points with now
+        private List<DateTimeParseResult> GetPrWithNow(MatchCollection matches, ExtractResult er, DateObject referenceDate)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Rename to ExtractNowAsDate or ParseNowAsDate.

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: I'd prefer Extract, and Parse later, as needed. Similar to the behaviour in the original method.

Comment (songwenhao1) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Ok, i will rename the function
 In the Parse stage of the  original method, DateParser couldn't parse the "now" for now is a datetime value, so return a pr about now maybe more simple

File: Specs/DateTime/English/DateTimeModel.json [hunk 15]
```diff
@@ -11060,5 +11060,55 @@
         }
       }
     ]
+  },
+  {
+    "Input": "I will do my work between now and November 15th",
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json]: Please add these same cases in DateTimeComplexCalendar too.

Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json]: Can you also add a case here and in ComplexCalendar like "Let's meet now". 
It should be recognized as a DateTime, not as a Date.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 16]
```diff
@@ -1262,42 +1262,57 @@ private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject refe
             var ret = new DateTimeResolutionResult();
 
             var er = this.config.DateExtractor.Extract(text, referenceDate);
+            var pr1 = new DateTimeParseResult();
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Why do you need to create two empty objects? Isn't it enough to declare the variable type?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 17]
```diff
@@ -1262,42 +1262,57 @@ private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject refe
...
-                er[1].Start -= this.config.TokenBeforeDate.Length;
             }
 
-            var match = this.config.WeekWithWeekDayRangeRegex.Match(text);
-            string weekPrefix = null;
-            if (match.Success)
-            {
-                weekPrefix = match.Groups["week"].ToString();
-            }
+            if (er.Count >= 2)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Why do we need this if now?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 18]
```diff
@@ -1325,6 +1340,36 @@ private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject refe
             return ret;
         }
 
+        // Parse entities that are made up by two time points with now
+        private DateTimeParseResult ParseNowAsDate(string text, DateObject referenceDate)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: What happens if there are two "now" in the sentence?
For example: "When I say now, I mean now!". Can you also add that as a spec case?

File: Specs/DateTime/English/DatePeriodExtractor.json [hunk 19]
```diff
@@ -3717,5 +3717,29 @@
...
+    "NotSupported": "python, javascript",
```
Comment (tellarin) [inline Specs/DateTime/English/DatePeriodExtractor.json]: If you added a Python implementation, why are these cases marked as not supported in it?

File: Specs/DateTime/English/DatePeriodExtractor.json [hunk 20]
```diff
@@ -3717,5 +3717,29 @@
...
+        "Text": "between now and november 15th",
+        "Type": "daterange",
+        "Start": 18,
+        "Length": 29
+      }
+    ]
+  },
+  {
+    "Input": "I have finished my work between Jan 22 and now",
+    "NotSupported": "python, javascript",
```
Comment (tellarin) [inline Specs/DateTime/English/DatePeriodExtractor.json]: And what's the key reason to not support in JS?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 18 above)
Comment (songwenhao1) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: We will match "between..and.." in stage Extract, so if just have two "now" and not have "between/from" , the text would not be extracted.
I will add the example as a spec case.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 17 above)
Comment (songwenhao1) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: In the former case ”er.count<2“, the value of er.count maybe changed in

er = this.config.DateExtractor.Extract(this.config.TokenBeforeDate + text, referenceDate);

so, i  judge the value again. if the value>=2, we should parse the er to pr, otherwise not needed

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 18 above)
Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: But the method usage is not clear. Even the code description.
What about a phrase like "Let's meet between now and Friday; not right now, ok?"?
Please also add it in specs.
